File: kotlin/bank-account/src/main/kotlin/BankAccount.kt
import java.util.concurrent.atomic.AtomicLong

class BankAccount {
    private var _b = AtomicLong(0L)
    var open = true
    val balance: Long
        get() {
            check(open)

            return _b.get()
        }

    fun adjustBalance(amount: Long) {
        check(open)
        _b.addAndGet(amount)
    }

    fun close() {
        open = false
    }
}

File: scala/bank-account/src/main/scala/BankAccount.scala
import java.util.concurrent.atomic.AtomicInteger

trait BankAccount {
  def closeAccount(): Unit
  def getBalance: Option[Int]
  def incrementBalance(increment: Int): Option[Int]
}
class BankAccountImpl() extends BankAccount {
  private var balance: Option[AtomicInteger] = Some(new AtomicInteger(0))
  def closeAccount(): Unit = balance = None
  def getBalance = balance.map(_.intValue())
  def incrementBalance(increment: Int) = balance.map(_.addAndGet(increment))
}
object Bank {
  def openAccount() = new BankAccountImpl()
}
